File: dataset/output/PERPLEXITY_llama-3-sonar-large-32k-chat/Task106_PERPLEXITY_llama-3-sonar-large-32k-chat.java
import java.math.BigInteger;
import java.security.SecureRandom;
import java.util.Scanner;

public class RSA {
    private BigInteger p, q, n, phi, e, d;

    public RSA(int bits) {
        SecureRandom random = new SecureRandom();
        p = new BigInteger(bits / 2, 100, random);
        q = new BigInteger(bits / 2, 100, random);
        n = p.multiply(q);
        phi = p.subtract(BigInteger.ONE).multiply(q.subtract(BigInteger.ONE));
        e = new BigInteger("65537"); // 2^16 + 1
        d = e.modInverse(phi);
    }

    public String encrypt(String message) {
        BigInteger msg = new BigInteger(message.getBytes());
        BigInteger encrypted = msg.modPow(e, n);
        return encrypted.toString();
    }

    public String decrypt(String encrypted) {
        BigInteger encryptedMsg = new BigInteger(encrypted);
        BigInteger decrypted = encryptedMsg.modPow(d, n);
        return new String(decrypted.toByteArray());
    }

    public static void main(String[] args) {
        RSA rsa = new RSA(1024);
        Scanner scanner = new Scanner(System.in);
        System.out.println("Enter a message:");
        String message = scanner.nextLine();
        String encrypted = rsa.encrypt(message);
        System.out.println("Encrypted: " + encrypted);
        String decrypted = rsa.decrypt(encrypted);
        System.out.println("Decrypted: " + decrypted);
    }
}

File: dataset/output-bk/PERPLEXITY_llama-3-sonar-large-32k-chat/Task106_PERPLEXITY_llama-3-sonar-large-32k-chat.cpp
#include <iostream>
#include <gmpxx.h>

void generate_keys(mpz_class& p, mpz_class& q, mpz_class& n, mpz_class& phi, mpz_class& e, mpz_class& d) {
    p = 61; // prime number
    q = 53; // prime number
    n = p * q;
    phi = (p - 1) * (q - 1);
    e = 17; // coprime with phi
    d = e.get_mod_inverse(phi);
}

void encrypt(const mpz_class& msg, mpz_class& encrypted, const mpz_class& e, const mpz_class& n) {
    encrypted = powm(msg, e, n);
}

void decrypt(const mpz_class& encrypted, mpz_class& decrypted, const mpz_class& d, const mpz_class& n) {
    decrypted = powm(encrypted, d, n);
}

int main() {
    mpz_class p, q, n, phi, e, d, msg, encrypted, decrypted;
    std::cout << "Enter a message (number): ";
    std::cin >> msg;
    generate_keys(p, q, n, phi, e, d);
    encrypt(msg, encrypted, e, n);
    std::cout << "Encrypted: " << encrypted << std::endl;
    decrypt(encrypted, decrypted, d, n);
    std::cout << "Decrypted: " << decrypted << std::endl;
    return 0;
}